test/e2e/upgrades/serviceaccount_admission_controller_migration.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrades
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/onsi/ginkgo"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	e2eauth "k8s.io/kubernetes/test/e2e/auth"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+)
+
+const (
+	podBeforeMigrationName = "pod-before-migration"
+	podAfterMigrationName  = "pod-after-migration"
+)
+
+// ServiceAccountAdmissionControllerMigrationTest test that a pod is functioning before and after
+// a cluster upgrade.
+type ServiceAccountAdmissionControllerMigrationTest struct {
+	pod *v1.Pod
+}
+
+// Name returns the tracking name of the test.
+func (ServiceAccountAdmissionControllerMigrationTest) Name() string {
+	return "[sig-auth] serviceaccount-admission-controller-migration"
+}
+
+// Setup creates pod-before-migration which has legacy service account token.
+func (t *ServiceAccountAdmissionControllerMigrationTest) Setup(f *framework.Framework) {
+	t.pod = createPod(f, podBeforeMigrationName)
+	inClusterClientMustWork(f, t.pod)
+}
+
+// Test waits for the upgrade to complete, and then verifies pod-before-migration
+// and pod-after-migration are able to make requests using in cluster config.
+func (t *ServiceAccountAdmissionControllerMigrationTest) Test(f *framework.Framework, done <-chan struct{}, upgrade UpgradeType) {
+	ginkgo.By("Waiting for upgrade to finish")
+	<-done
+
+	ginkgo.By("Starting post-upgrade check")
+	podBeforeMigration, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Get(context.TODO(), podBeforeMigrationName, metav1.GetOptions{})
+	framework.ExpectNoError(err)
+	if podBeforeMigration.GetUID() != t.pod.GetUID() {
+		framework.Failf("Pod %q GetUID() = %q, want %q.", podBeforeMigration.Name, podBeforeMigration.GetUID(), t.pod.GetUID())
+	}
+	if podBeforeMigration.Status.ContainerStatuses[0].RestartCount != 0 {
+		framework.Failf("Pod %q RestartCount = %d, want 0.", podBeforeMigration.Name, podBeforeMigration.Status.ContainerStatuses[0].RestartCount)
+	}
+	if len(podBeforeMigration.Spec.Volumes) != 1 || podBeforeMigration.Spec.Volumes[0].Secret == nil {
+		framework.Failf("Pod %q Volumes[0].Secret = nil, want non-nil.", podBeforeMigration.Name)
+	}
```

Comment:
this check will fail if the "from" version this test suite is run with already had the bound tokens enabled... we want this test to be robust into the future

Reply:
I also expected a call to `inClusterClientMustWork(f, podBeforeMigration)` here, passing in the time of the upgrade so we could verify correctness of the pod after upgrade

Reply:
done

Reply:
inClusterClientMustWork looks good. drop the `if len(podBeforeMigration.Spec.Volumes) != 1 || podBeforeMigration.Spec.Volumes[0].Secret == nil {` check... the only thing we really care about is that the pod created before the upgrade still works after the upgrade

Reply:
done